model/template.go

```diff
@@ -77,11 +77,150 @@ func fillInTemplate(db *gorm.DB, templateModel *TemplateModel) {
 	if catalog != nil {
 		templateModel.Catalog = catalog.Name
 	}
+
 	templateModel.Categories = lookupTemplateCategories(db, templateModel.ID)
 	templateModel.Labels = lookupTemplateLabels(db, templateModel.ID)
 	templateModel.Versions = lookupVersions(db, templateModel.ID)
 }
 
+func templateCategoryMap(db *gorm.DB, templateIDList []int) map[int][]string {
+	categoriesQuery := `
+	SELECT template_id, category_id, name
+	FROM catalog_template_category tc
+	JOIN catalog_category c ON (tc.category_id = c.id)
+	WHERE tc.template_id IN ( ? )`
+
+	catagoryAndTemplateList := []CategoryAndTemplate{}
+	db.Raw(categoriesQuery, templateIDList).Find(&catagoryAndTemplateList)
+
+	// make map of template (key) to category name (value)
+
+	var catagoryAndTemplateMap map[int][]string
+	catagoryAndTemplateMap = make(map[int][]string)
+	for _, catagoryAndTemplate := range catagoryAndTemplateList {
+
+		catagoryAndTemplateMap[catagoryAndTemplate.TemplateID] = append(catagoryAndTemplateMap[catagoryAndTemplate.TemplateID], catagoryAndTemplate.Name)
+
+	}
+
+	return catagoryAndTemplateMap
+}
+
+func templateLabelMap(db *gorm.DB, templateIDList []int) map[int]map[string]string {
+	labelsQuery := "SELECT template_id, `key`, value FROM catalog_label cl WHERE cl.template_id IN ( ? )"
+
+	var labelAndTemplateList []TemplateLabelModel
+	db.Raw(labelsQuery, templateIDList).Find(&labelAndTemplateList)
+
+	var labelAndTemplateMap map[int]map[string]string
+	labelAndTemplateMap = make(map[int]map[string]string)
+
+	for _, labelAndTemplate := range labelAndTemplateList {
+
+		if _, ok := labelAndTemplateMap[int(labelAndTemplate.TemplateId)]; !ok {
+			labels := map[string]string{labelAndTemplate.Key: labelAndTemplate.Value}
+			labelAndTemplateMap[int(labelAndTemplate.TemplateId)] = labels
+		} else {
+			labelAndTemplateMap[int(labelAndTemplate.TemplateId)][labelAndTemplate.Key] = labelAndTemplate.Value
+		}
+	}
+
+	return labelAndTemplateMap
+}
+
+func templateVersionMap(db *gorm.DB, templateIDList []int) map[int][]Version {
+	var versionList []VersionModel
+
+	// all versions with list of template IDs
+	versionsQuery := `
+	SELECT *
+	FROM catalog_version
+	WHERE catalog_version.template_id IN ( ? )`
+	db.Raw(versionsQuery, templateIDList).Find(&versionList)
+
+	// look up version based on version id
+	versionMap := map[uint]VersionModel{}
+	var versionIDs []int
+
+	for _, version := range versionList {
+
+		versionIDs = append(versionIDs, int(version.ID))
+
+		versionMap[version.ID] = version
+	}
+
+	var versionLabelList []VersionLabelModel
+
+	versionLabelsQuery := `
+	SELECT *
+	FROM catalog_version_label
+	WHERE catalog_version_label.version_id IN (?)`
+
+	db.Raw(versionLabelsQuery, versionIDs).Find(&versionLabelList)
+
+	for _, label := range versionLabelList {
+
+		if versionMap[label.VersionId].Labels == nil {
+			version := versionMap[label.VersionId]
+			version.Labels = make(map[string]string)
+
+			versionMap[label.VersionId] = version
```

Comment:
This is redundant, `versionMap[label.VersionId]` is already `version` because it was set from that 2 lines ago...